kubernetes/updater/pkg/img/validator.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package img
+
+import (
+	"context"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/docker/distribution/reference"
+	"github.com/opencontainers/go-digest"
+)
+
+type NamedTaggedDigested interface {
+	reference.Named
+	reference.Tagged
+	reference.Digested
+}
+
+type Validator interface {
+	Name() string
+	ValidateAndResolveDigest(ctx context.Context, image reference.NamedTagged) ([]NamedTaggedDigested, error)
+}
+
+func AtLeastOneValid(ctx context.Context, image reference.NamedTagged, validators ...Validator) (bool, []NamedTaggedDigested) {
+	var resolvedImages []NamedTaggedDigested
+	var err error
+	log := ctrllog.FromContext(ctx).V(1)
+	for _, validator := range validators {
+		resolvedImages, err = validator.ValidateAndResolveDigest(ctx, image)
+		log.Info("Image validation result", "img", image, "validator", validator.Name(), "err", err)
```

Comment:
```suggestion
		log.Error(err,"Image validation result", "img", image, "validator", validator.Name())
```

Reply:
I'm not sure how `logr.Error()` behaves when called with a `nil` error, will have to test

Reply:
Logging in the validation was messy, and I wanted to log errors only when all validators failed. Some validators failing is a legitimate outcome (e.g. multiple keys or signature methods) and should not send scary logs to the user. 

Now AtLeastOne() returns an aggregate containing the validation failures of all validators if none matched. I also discovered the `trace.Trust` error which is a better fit than `trace.CompareFailed`.

Can you check if the new behaviour looks saner?

Reply:
:nail_care: 
Right now we'll print twice for a successful check.
If we move the log.Info to only happen after the `if err == nil` check, we would print once per validator (with success or an error).